### PR TITLE
add covStub to file object when includeUntested is set

### PR DESCRIPTION
**Pull request by glan**
This allows downstream consumers to produces a report with includes uncovered files.

### CODE REVIEW

**Comment by SBoudrias** [comment]
Can you provide a unit test?

**Comment by glan** [comment]
@SBoudrias Let me know if that looks alright not not.

File: test/main.js [hunk 1]
```diff
@@ -73,6 +74,35 @@ describe('gulp-istanbul', function () {
...
+
+    it('adds covStub to untested files', function (done) {
+      var COV_VAR = 'untestedCovVar';
+      var covStubs = {};
+      gulp.src([ 'test/fixtures/lib/*.js' ])
+        .pipe(istanbul({
+            coverageVariable: COV_VAR,
+            includeUntested: true
+        }))
+        .pipe(es.through(function (file) {
```
Comment (SBoudrias) [inline test/main.js]: You can use `through2` here and that won't require a new dependency.

File: test/main.js [hunk 2]
```diff
@@ -73,6 +74,35 @@ describe('gulp-istanbul', function () {
...
+          covStubs[file.path] = file.covStub;
+          this.push(file);
+        }))
+        .on('end', function () {
+          var filePaths = Object.keys(covStubs);
+          assert(filePaths.length === 2, '2 files with covStub');
+          filePaths.forEach(function (filePath) {
+            assert(covStubs[filePath].path === filePath, 'covStub has path');
+            assert(covStubs[filePath].hasOwnProperty('s'), 'covStub has s');
```
Comment (SBoudrias) [inline test/main.js]: We don't actually care about the content of the covStub property, that's an external implementation details. As soon as the module feature changes our test will fail - that's not what we want.

**Comment by SBoudrias** [comment]
Also... Wouldn't this actually only works if you `includeUntested`? We never export `covStub` for covered file if we don't include the `includeUntested` flag

**Comment by glan** [comment]
@SBoudrias yeap it's only intended to be used with includeUntested. Basically it's purpose is to create a base line stub coverage stream which gets merged in with actual coverage stream later. I'll get `es.through` swapped out with `through2` I'm just less familiar with that one.

**Comment by SBoudrias** [comment]
Can you explain exactly how you would use this code? With code example of an implementation using it?

I get the feeling this is inconsistent and actually a bit hacky.

**Comment by glan** [comment]
Here's the idea:

```js
var gulpIstanbul = require('gulp-istanbul');
var browserifyIstanbul = require('bowserify-istanbul');

// 1. stub coverage data for all src files:
var stubCoverageStream = gulp.src(['./src/**/*.js'])
    .pipe(gulpIstanbul({
        includeUntested: true
    }));

// 2. run tests and collect coverage as a stream:
var testCoverageStream = gulp.src(['./test/specs/**/*.js'], {
        read: false
    })
    .pipe(browserify({
        debug: true,
        transform: [browserifyIstanbul]
    }))
    .pipe(karma({
        frameworks: ['jasmine'],
        reporters: ['coverage'],
        browsers: ['PhantomJS'],
        action: 'run'
    }));

// 3. collect coverage data from both streams and produce report
es.merge(testCoverage, blankCoverage)
    .pipe(customIstanbulReporter);
```

**Comment by SBoudrias** [comment]
What are `testCoverage` and `blankCoverage` streams?

**Comment by glan** [comment]
@SBoudrias sorry they should have been named `stubCoverageStream` and `testCoverageStream`. These are vinyl objects either containing coverage data in their `content` or attached in `covStub` property.

**Comment by SBoudrias** [comment]
I think what I'd like is to update the untested hack to simply force require the test file. Then everything will be 100% handled by Istanbul without weird hack on our parts.

**Comment by glan** [comment]
I was also thinking that we could do this without updating gulp-istanbul by pipping the output into a process which requires each covered file and returns the stub coverage data. Let me write an example of that.

**Comment by glan** [comment]
@SBoudrias probably better not to make this change and just do:

 ```js
var stubCoverageStream
  .pipe(es.through(function (file) {
     this.push(global[coverageVar][file.path]);
}));
```

or

```js
var stubCoverageStream
  .pipe(es.through(function (file) {
    var instrumentedSrc = file.contents.toString();
    var covStubRE = /\{.*"path".*"fnMap".*"statementMap".*"branchMap".*\}/g;
    var covStubMatch = covStubRE.exec(instrumentedSrc);
    if (covStubMatch !== null) {
      this.push(JSON.parse(covStubMatch[0]));
    }
}));

**Pull request closed without merging**